### PR TITLE
Feature/postpayment

### PR DESCRIPTION
**Pull request by RvdK**
Allows you to create add a transaction to the bunq account

### CODE REVIEW

**Comment by joskuijpers** [comment]
Could you rebase on develop and push? Then travis should be nicer.

**Comment by RvdK** [comment]
rebase did not help

**Comment by joskuijpers** [comment]
Rebase again, after the develop build finished. It should work... (it must :'( )

**Comment by RvdK** [comment]
merged the extended testing with this pr as they are related

**Review by joskuijpers — APPROVED** [review]

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java [hunk 1]
```diff
@@ -144,4 +153,100 @@ public void testExceptionConfirmationNotFuture() {
 
         assertTrue(bank.confirmException(thr) == thr);
     }
+
+    @Test
+    public void testGetUsers() throws ExecutionException, InterruptedException {
+        CompletableFuture<Void> future;
```
Comment (MatthijsKok) [inline BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java]: Unused, please remove. Please think about putting other variables in BeforeClass or Before methods to remove a lot of boilerplate.

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java [hunk 2]
```diff
@@ -144,4 +153,100 @@ public void testExceptionConfirmationNotFuture() {
...
+        users = bank.listUsers().get();
+        Party firstUser = users.get(0);
+        assertEquals(firstUser.getId(), 2002);
+        assertEquals(firstUser.getName(), "Overton Onderlinge Waarborgmaatschappij");
+    }
+
+
+    @Test
+    public void testGetAccounts() throws ExecutionException, InterruptedException {
+        CompletableFuture<Void> future;
```
Comment (MatthijsKok) [inline BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java]: Unused, please remove.

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java [hunk 3]
```diff
@@ -144,4 +153,100 @@ public void testExceptionConfirmationNotFuture() {
...
+        Account ac = accounts.get(0);
+        assertEquals(ac.getParty().getId(), 2002);
+        assertEquals(ac.getIban(), "NL05BUNQ9900019989" +
+                "");
+        assertEquals(ac.getId(), 2021);
+    }
+
+    @Test
+    public void testListTransactions() throws ExecutionException, InterruptedException {
+        CompletableFuture<Void> future;
```
Comment (MatthijsKok) [inline BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java]: Unused, please remove.

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java [hunk 4]
```diff
@@ -144,4 +153,100 @@ public void testExceptionConfirmationNotFuture() {
...
+        Transaction t = transactions.get(transactions.size() - 2);
+        assertEquals(t.getId(), 14191);
+        assertEquals(t.getDescription(), "invoice 688");
+        assertEquals(t.getAcount().getId(), 2021);
+        assertEquals(t.getCounterAccount().getIban(), "NL28BUNQ9900000951");
+    }
+
+    @Test
+    public void testPostTransaction() throws ExecutionException, InterruptedException {
+        CompletableFuture<Void> future;
```
Comment (MatthijsKok) [inline BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java]: Unused, please remove.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java [hunk 5]
```diff
@@ -13,6 +13,7 @@
 
 import java.io.IOException;
 import java.util.ArrayList;
+
```
Comment (MatthijsKok) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java]: Again, please check your staged files and changes before committing. This is easily avoidable. Please revert

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java [hunk 6]
```diff
@@ -91,6 +92,7 @@ public void retrieveRecentTransactions() {
                         Log.e("CRYPTO", e.getMessage());
                     }
 
+
```
Comment (MatthijsKok) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/activities/RecentTransactionsActivity.java]: Please revert.

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java [hunk 7]
```diff
@@ -144,4 +153,100 @@ public void testExceptionConfirmationNotFuture() {
 
         assertTrue(bank.confirmException(thr) == thr);
     }
+
+    @Test
+    public void testGetUsers() throws ExecutionException, InterruptedException {
```
Comment (MatthijsKok) [inline BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java:159]: Please also add tests for bad code paths and any exceptions that might be thrown.

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java [hunk 8]
```diff
@@ -144,4 +153,100 @@ public void testExceptionConfirmationNotFuture() {
...
+        bank.createSession().get();
+
+        BunqParty party = new BunqParty("Overton Onderlinge Waarborgmaatschappij", 2002);
+        BunqAccount account = new BunqAccount("NL05BUNQ9900019989", 2021, party);
+
+        BunqParty cp = new BunqParty("Branda Monoghan", -1);
+        BunqAccount ca = new BunqAccount("NL77BUNQ9900016947", -1, cp);
+        BunqTransaction pay = new BunqTransaction(-0.01f, account, ca, Currency.getInstance("EUR"), "end tot end Bunqtest");
+
+        assertTrue(bank.sendTransaction(pay).get());
```
Comment (MatthijsKok) [inline BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java]: Please also test that the balances are changed, the transaction shows up when you get the transaction list, etc...

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java [hunk 9]
```diff
@@ -144,4 +153,100 @@ public void testExceptionConfirmationNotFuture() {
...
+        env.setBankApiKey("REAL_BUNQ_API_KEY");
+
+        bank = new BankFactory(env).create();
+        bank.createSession().get();
+
+        BunqParty party = new BunqParty("Overton Onderlinge Waarborgmaatschappij", 2002);
+        BunqAccount account = new BunqAccount("NL05BUNQ9900019989", 2021, party);
+
+        List<Transaction> transactions = bank.listTransactions(account).get();
+        Transaction t = transactions.get(transactions.size() - 2);
```
Comment (MatthijsKok) [inline BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java]: Please test that multiple transactions exist, since it is a list.

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java [hunk 10]
```diff
@@ -144,4 +153,100 @@ public void testExceptionConfirmationNotFuture() {
...
+        List<Account> accounts = bank.listAccount(new BunqParty("Overton Onderlinge Waarborgmaatschappij", 2002)).get();
+        Account ac = accounts.get(0);
+        assertEquals(ac.getParty().getId(), 2002);
+        assertEquals(ac.getIban(), "NL05BUNQ9900019989" +
+                "");
+        assertEquals(ac.getId(), 2021);
+    }
+
+    @Test
+    public void testListTransactions() throws ExecutionException, InterruptedException {
```
Comment (MatthijsKok) [inline BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java]: Please also add a test that returns an empty list.

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java [hunk 11]
```diff
@@ -144,4 +153,100 @@ public void testExceptionConfirmationNotFuture() {
...
+        bank.createSession().get();
+        users = bank.listUsers().get();
+        Party firstUser = users.get(0);
+        assertEquals(firstUser.getId(), 2002);
+        assertEquals(firstUser.getName(), "Overton Onderlinge Waarborgmaatschappij");
+    }
+
+
+    @Test
+    public void testGetAccounts() throws ExecutionException, InterruptedException {
```
Comment (MatthijsKok) [inline BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java:182]: Please test that multiple accounts exist, since it is a list. Also add a test for empty list etc..

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java [hunk 12]
```diff
@@ -162,6 +161,25 @@ private Void loadOrCreateClientKeys() {
         });
     }
 
+    @Override
+    public CompletableFuture<Boolean> sendTransaction(Transaction transaction) {
+        BunqTransaction tran = (BunqTransaction) transaction;
```
Comment (MatthijsKok) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java:166]: please check instanceof before casting.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java [hunk 13]
```diff
@@ -162,6 +161,25 @@ private Void loadOrCreateClientKeys() {
...
+        CompletableFuture<PaymentService.PostResponse> future;
+        PaymentService service;
+
+        service = retrofit.create(PaymentService.class);
+
+        future = service.createPayment(tran.convertToRequest(), tran.getAcount().getParty().getId(), tran.getAcount().getId());
+        return future.thenApply((PaymentService.PostResponse response) -> {
```
Comment (MatthijsKok) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java:173]: Please change PaymentService.PostResponse into an import.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTransaction.java [hunk 14]
```diff
@@ -22,15 +24,35 @@
     private Account account;
 
 
-    BunqTransaction(PaymentService.ListResponse.Payment payment, Account account) {
+    public BunqTransaction(PaymentService.ListResponse.Payment payment, Account account) {
```
Comment (MatthijsKok) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTransaction.java:27]: Please change PaymentService.ListReponse.Payment into an import.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTransaction.java [hunk 15]
```diff
@@ -64,7 +86,30 @@ public String getDescription() {
...
+    public void setId(int id) {
+        this.id = id;
+    }
+
+    @Override
     public String toString() {
         return "<Payment desc:'" + getDescription() + "'>";
     }
+
+    public PaymentService.PostRequest convertToRequest() {
```
Comment (MatthijsKok) [inline BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTransaction.java:102]: Please change PaymentService.PostRequest into an import, and others below as well.

**Review by MatthijsKok — CHANGES_REQUESTED** [review]
Some additional changes. code is a good addition.

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java (hunk 8 above)
Comment (RvdK) [reply BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java]: I do not agree because bunq approves the transaction when returning true which is enough

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java (hunk 13 above)
Comment (joskuijpers) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java:173]: Don't, as PostResponse is not a unique class name: it is an inner name of any API that has a post response object. It is also not clear when reading the code what "PostResponse" refers to: it is now.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTransaction.java (hunk 14 above)
Comment (joskuijpers) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTransaction.java:27]: Again, see previous comment

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTransaction.java (hunk 15 above)
Comment (joskuijpers) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTransaction.java:102]: Again, don't. See previous comments.

Comment (RvdK) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTransaction.java:102]: Disagree we have many post requests (posssibly for each service) and that would decrease readablialty

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java (hunk 9 above)
Comment (joskuijpers) [reply BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java]: The list size constantly changes and can't be independently verified. Only thing that can be done is that the size is at least 3. But if that is not the case, get() will throw an exception and thus also fail the test.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java (hunk 12 above)
Comment (RvdK) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java:166]: instance of is noet needed here since transaction is extended by bunqtransaction so this will always succeed

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java (hunk 11 above)
Comment (RvdK) [reply BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java:182]: We can't this is a live test and we can not add accounts

File: BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java (hunk 7 above)
Comment (MatthijsKok) [reply BankChain/app/src/androidTest/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqTest.java:159]: We agreed to not test any bad code paths, since we are measured by line coverage instead of branch coverage.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java (hunk 12 above)
Comment (MatthijsKok) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java:166]: Agreed.

File: BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java (hunk 13 above)
Comment (MatthijsKok) [reply BankChain/app/src/main/java/nl/tudelft/ewi/ds/bankchain/bank/bunq/BunqBank.java:173]: Agreed.

**Review by MatthijsKok — APPROVED** [review]
LGTM with changes
